src/osd/ReplicatedBackend.cc

```diff
@@ -2104,6 +2192,10 @@ int ReplicatedBackend::build_push_op(const ObjectRecoveryInfo &recovery_info,
     new_progress.data_complete = true;
     if (stat)
       stat->num_objects_recovered++;
+  } else {
+    // If omap is not changed, we need recovery omap when recovery cannot be completed once
+    if (progress.first && progress.omap_complete)
+      new_progress.omap_complete = false;
```

Comment:
the implementation is here. If omap is not change and recovery cannot finish in one PushOp, then recovery, then transmit omap again in next PushOp. I think this is the simplest figuring out for the problem, right?